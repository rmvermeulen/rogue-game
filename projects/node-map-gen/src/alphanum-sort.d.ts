declare module 'alphanum-sort' {
  // tslint:disable
  const sort: <T extends any[]>(input: T) => T;
  export = sort;
}
