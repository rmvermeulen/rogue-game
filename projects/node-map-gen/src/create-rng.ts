import * as Chance from 'chance';

export const createRNG = (seed?: number) => new Chance(seed);

export const rng = createRNG();
