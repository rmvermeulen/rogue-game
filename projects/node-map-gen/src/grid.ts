import { times, objOf, whereEq } from 'ramda';

const createObjects = times(objOf('id'));

export interface IGridOptions {
  width: number;
  height: number;
}

export interface ICell {
  id: number;
  x: number;
  y: number;
  room: number;
}

export class Grid {
  public readonly width;
  public readonly height;
  public readonly cells: ICell[];
  private constructor({ width, height }: IGridOptions) {
    this.width = width;
    this.height = height;

    this.cells = createObjects(width * height).map(cell =>
      Object.assign(cell, {
        x: cell.id % width,
        y: Math.floor(cell.id / width),
        room: 0,
      }),
    );
  }
  static create(options: IGridOptions) {
    return new Grid(options);
  }

  cellAt(x: number, y: number): ICell | undefined {
    return this.cells.find(whereEq({ x, y }));
  }

  row(y: number): ICell[] {
    return this.cells.filter(whereEq({ y }));
  }

  column(x: number): ICell[] {
    return this.cells.filter(whereEq({ x }));
  }

  listRooms(): number[] {
    return Array.from(
      this.cells.reduce((ids, cell) => {
        ids.add(cell.room);
        return ids;
      }, new Set<number>()),
    );
  }

  display(): string {
    return '';
  }
}
