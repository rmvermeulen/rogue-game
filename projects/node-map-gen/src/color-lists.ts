import chalk from 'chalk';
import { rng } from './create-rng';

const colors = [
  chalk.red,
  chalk.green,
  chalk.blue,
  chalk.yellow,
  chalk.magenta,
  chalk.cyan,
  chalk.bgRed.black,
  chalk.bgGreen.black,
  chalk.bgBlue.black,
  chalk.bgYellow.black,
  chalk.bgMagenta.black,
  chalk.bgCyan.black,
];

export const someColorsLength = colors.length;

export const someColors = () => rng.shuffle(colors);

const brights = [
  chalk.redBright,
  chalk.greenBright,
  chalk.blueBright,
  chalk.yellowBright,
  chalk.magentaBright,
  chalk.cyanBright,
  chalk.bgRedBright.black,
  chalk.bgGreenBright.black,
  chalk.bgBlueBright.black,
  chalk.bgYellowBright.black,
  chalk.bgMagentaBright.black,
  chalk.bgCyanBright.black,
];

export const manyColorsLength = colors.length + brights.length;

export const manyColors = () => rng.shuffle(colors.concat(brights));
