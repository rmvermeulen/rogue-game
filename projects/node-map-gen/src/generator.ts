import { assert } from 'chai';
import {
  __,
  addIndex,
  assoc,
  chain,
  compose,
  head,
  map,
  merge,
  prop,
  sortBy,
} from 'ramda';
import { ICell, IGridOptions } from './grid';
import { Pool } from './pool';
import {
  Candidate,
  createCell,
  findCandidatesFrom,
  gridFactory,
  manhattan,
  weightedPick,
} from './utils';

const applyRoomIds = addIndex<ICell[], ICell[][], ICell[]>(chain)(
  (cells: ICell[], roomId: number): ICell[] => {
    assert.isArray(cells, 'applyRoomIds');

    return cells.map(merge(__, { roomId }));
  },
);

type Scored<T> = [T, number];
const pickChampion = (input: Array<Scored<Candidate>>): Candidate => {
  if (input.length === 0) {
    return undefined;
  }

  const [champ] = input.reduce(
    (
      candidate: Scored<Candidate>,
      champion: Scored<Candidate>,
    ): Scored<Candidate> => (champion[1] < candidate[1] ? champion : candidate),
  );

  return champ;
};

const scoreCandidateForRoom = (room: ICell[]) => (
  candidate: Candidate,
): Scored<Candidate> => {
  const totalDistance = room.reduce(
    (total, cell) =>
      total +
      (Math.sqrt((candidate.x - cell.y) ** 2) +
        Math.sqrt(candidate.y - cell.y) ** 2),
    0,
  );

  return [candidate, totalDistance];
};

const sortByScore: (
  cs: Array<Scored<Candidate>>,
) => Array<Scored<Candidate>> = sortBy(prop(1));

const preferCloser = (
  currentRoom: ICell[],
  candidates: Candidate[],
): Candidate =>
  // tslint:disable-next-line: no-unsafe-any
  compose(
    head,
    weightedPick,
    sortByScore,
    map(scoreCandidateForRoom(currentRoom)),
  )(candidates);

const pickClosest = (
  currentRoom: ICell[],
  candidates: Candidate[],
): Candidate =>
  compose<Candidate[], Array<Scored<Candidate>>, Candidate>(
    pickChampion,
    map(scoreCandidateForRoom(currentRoom)),
  )(candidates);

// tslint:disable-next-line: no-empty-interface
export interface IGeneratorOptions extends IGridOptions {
  pickMethod?: 'random' | 'closest' | 'prefer closer';
}

export const generateCells = (options: IGeneratorOptions): ICell[] => {
  const cells = gridFactory(
    options.width,
    options.height,
    compose(
      assoc('roomId', -1),
      createCell,
    ),
  );
  const pool = new Pool(cells, options.rng);
  const idOf = prop('id');
  pool.useEq((a, b) => idOf(a) === idOf(b));
  // pick a starting cell for each room
  // must be unique (handled by pool)
  // must not be locked in on 4 sides
  const initialCells = pool.takeN(options.roomCount, (cell, set) => {
    let closedSides = 0;
    for (const other of set) {
      if (manhattan(cell, other) === 1) {
        closedSides += 1;
      }
      if (closedSides >= 4) {
        return false;
      }
    }

    return true;
  });
  assert.lengthOf(
    initialCells,
    options.roomCount,
    'Must have an initial cell for each room',
  );
  const assignedCells = [...initialCells];
  const rooms: ICell[][] = initialCells.map(cell => [cell]);
  // remaining cells are the pool
  let safetyLimit = 10000;
  // for each room

  while (safetyLimit > 0 && pool.size > 0) {
    safetyLimit -= 1;
    for (let i = 0; i < options.roomCount; i += 1) {
      // if pool is empty break;
      if (pool.size === 0) {
        break;
      }
      // find cells adjacent to cell in current room
      // NOTE: maybe use pool.take[N/Many] with constraints

      const candidates = findCandidatesFrom(pool.ref())(rooms[i], true);
      if (candidates.length === 0) {
        continue;
      }

      let picked: Candidate | undefined;
      switch (options.pickMethod) {
        case 'random':
          picked = options.rng.pickone(candidates);
          break;

        case 'closest':
          picked = pickClosest(rooms[i], candidates);
          break;

        default:
        case 'prefer closer':
          picked = preferCloser(rooms[i], candidates);
      }

      if (picked === undefined) {
        continue;
      }

      // remove from pool
      pool.remove([picked]);
      // add cell to list for the current room
      rooms[i].push(picked);
      assignedCells.push(picked);
    }
  }

  assert(safetyLimit, 'Too many loops');

  return compose(
    sortBy(prop('id')),
    applyRoomIds,
  )(rooms);
};
