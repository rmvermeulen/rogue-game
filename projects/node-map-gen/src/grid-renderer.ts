import { assert } from 'chai';
import * as os from 'os';
import {
  aperture,
  compose,
  defaultTo,
  length,
  lensPath,
  map,
  path,
  pluck,
  repeat,
  set,
  sort,
  subtract,
  test,
  toString,
  unnest,
  when,
  zip,
} from 'ramda';
import {
  manyColors,
  manyColorsLength,
  someColors,
  someColorsLength,
} from './color-lists';
import { Grid, ICell } from './grid';
import { generate } from './utils';

export const renderSimple = (grid: Grid, useColors: boolean = true): string => {
  if (grid.listRooms().length > manyColorsLength) {
    // tslint:disable-next-line: no-parameter-reassignment
    useColors = false;
  }
  const colors = useColors
    ? grid.cells.length <= someColorsLength
      ? someColors()
      : manyColors()
    : [];

  const columnWidths = repeat(1, grid.width);

  generate(grid.height, (rowIndex: number) => {
    const row = grid.row(rowIndex);
    const roomStrWidths = compose<ICell[], number[], number[]>(
      map(
        compose<number, string, number>(
          length,
          toString,
        ),
      ),
      pluck<ICell, 'roomId'>('roomId'),
    )(row);
    // update column
    roomStrWidths.forEach((width, column) => {
      columnWidths[column] = Math.max(width, columnWidths[column]);
    });
  });

  return generate<string>(grid.height, (rowIndex: number) =>
    grid
      .row(rowIndex)
      .map((cell: ICell, index) => {
        let roomIdStr = cell.roomId.toString();
        const roomIdStrCharLength = roomIdStr.length;
        if (useColors) {
          const colorize = colors[cell.roomId % colors.length];
          roomIdStr = colorize(roomIdStr);
        }
        const width = columnWidths[index];

        // cannot use String.padEnd because of color in text
        return ' '.repeat(width - roomIdStrCharLength) + roomIdStr;
      })
      .join(' '),
  ).join(os.EOL);
};

// const stringifyLine = (
//   values: number[],
// ): { text: string; columnWidths: number[] } => {
//   return compose<number[], string[], { text: string; columnWidths: number[] }>(
//     applySpec({
//       text: (ss: string[]) => `| ${ss.join(' | ')} |`,
//       columnWidths: pluck('length'),
//     }),
//     map(String),
//   )(values);
// };

/** render the grid into a string
 * example output:
 * +----+----+---+----+-----+
 * | 10   10 | 2   2  | 200 |
 * +----+    +   +    +-----+
 * | 0  | 10 | 2   2  | 10  |
 * +    +    +   +    +-----+
 * | 0  | 10 | 2   2  | 0   |
 * +    +    +   +    +     +
 * | 0  | 10 | 2   2  | 0   |
 * +    +----+---+----+-----+
 * | 0    0    0 | 10   10  |
 * +--- +----+---+----+-----+
 *
 * room 0 size=9 cells=4,5,10,14,15,19,20,21,22
 * room 1 size=8 cells=0,1,6,9,11,16,23,24
 * room 2 size=8 cells=2,3,7,8,12,13,17,18
 */
// tslint:disable-next-line: max-func-body-length
export const render = (grid: Grid, useColors: boolean = true): string => {
  type WallStr = '|' | '+' | '-' | ' ';

  const columnWidths: number[] = generate(grid.width, () => 1);

  const mapLinesA = generate(grid.height, y => {
    const row = grid.row(y);
    const roomIds = pluck('roomId', row);
    generate(grid.width, column => {
      const tokenWidth = String(roomIds[column]).length;
      columnWidths[column] = Math.max(columnWidths[column], tokenWidth);
    });
    const walls: WallStr[] = [];
    for (const [a, b] of aperture(2, roomIds)) {
      const nextWall: WallStr = a === b ? ' ' : '|';
      walls.push(nextWall);
    }
    const [first, ...rest] = roomIds;

    return ['|', first, ...unnest(zip(walls, rest)), '|'];
  });

  const columnWalls: WallStr[][] = repeat([], grid.width);
  generate(grid.width, x => {
    const column = grid.column(x);
    const walls = columnWalls[x];
    walls.push('-');
    for (const [a, b] of aperture(2, pluck('roomId', column))) {
      const nextWall: WallStr = a === b ? ' ' : '-';
      walls.push(nextWall);
    }
    walls.push('-');
  });

  const topLine: WallStr[] = mapLinesA[0].map((value: number | WallStr) =>
    typeof value === 'number' ? '-' : value === '|' ? '+' : '-',
  );
  const lastMapLine = mapLinesA[grid.height - 1];
  const bottomLine: WallStr[] = lastMapLine.map((value: number | WallStr) =>
    typeof value === 'number' ? '-' : value === '|' ? '+' : '-',
  );
  const mapLinesB = [
    topLine,
    ...unnest(
      [...aperture(2, mapLinesA), [lastMapLine, undefined]].map(([a, b]) => {
        if (b === undefined) {
          return [a, bottomLine];
        }
        const walls = generate(a.length, x => {
          const ax = a[x];
          const bx = b[x];

          if (typeof ax === 'number') {
            return ax === bx ? ' ' : '-';
          }

          return ax === bx ? '|' : '+';
        });

        return [a, walls];
      }),
    ),
  ];

  assert.lengthOf(mapLinesB, grid.height * 2 + 1);

  let updated = mapLinesB;

  generate(grid.height + 1, gy => {
    generate(grid.width + 1, gx => {
      const x = gx * 2;
      const y = gy * 2;
      // const cornerPart = mapLinesB[y * 2][x * 2];
      const [n, s, e, w] = [[y - 1, x], [y + 1, x], [y, x - 1], [y, x + 1]]
        // tslint:disable-next-line: no-any no-unsafe-any
        .map(p => path(p as any[], mapLinesB))
        .map(defaultTo(' '));
      const assignField = (char: string) => {
        // tslint:disable-next-line: no-any
        const modify = set(lensPath([y, x]), char) as any;
        // tslint:disable-next-line: no-unsafe-any
        updated = modify(updated);
      };
      if (n === ' ' && s === ' ' && e === ' ' && w === ' ') {
        return assignField(' ');
      }
      if (e === ' ' && w === ' ') {
        return assignField('|');
      }
      if (n === ' ' && s === ' ') {
        return assignField('-');
      }

      return assignField('+');
    });
  });

  const lineFragments = updated.map((chunks, y) => {
    const newLine = chunks.map(String).map((chunk, x) => {
      const isRoomCenterX = x % 2 === 1;
      if (!isRoomCenterX) {
        return chunk;
      }
      const isRoomCenterY = y % 2 === 1;
      const width = defaultTo(1, columnWidths[(x - 1) / 2]);

      const padded = isRoomCenterY
        ? ` ${chunk.padStart(width)} `
        : chunk.repeat(width + 2);
      assert.lengthOf(padded, width + 2);

      return padded;
    });

    return newLine.join('');
  });

  const rooms = sort(subtract, grid.listRooms())
    .map(id => grid.findRoomById(id))
    .filter(Boolean);

  const roomFragments = rooms.map(
    ({ id, size, cells }) => `room ${id} size=${size} cells=${cells}`,
  );

  const renderedString = lineFragments.join(os.EOL);

  if (useColors) {
    const colors =
      grid.cells.length <= someColorsLength ? someColors() : manyColors();
    const getColorizer = (n: number) => colors[n % colors.length];

    return [
      renderedString
        .split(' ')
        .map(
          when(test(/^\d+$/), (str: string) => getColorizer(Number(str))(str)),
        )
        .join(' '),
      '',
      ...roomFragments,
    ].join(os.EOL);
  }

  return [renderedString, '', ...roomFragments].join(os.EOL);
};
