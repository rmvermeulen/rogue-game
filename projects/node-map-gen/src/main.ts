import { Grid } from './grid';

export const gridFactory = Grid.create;
