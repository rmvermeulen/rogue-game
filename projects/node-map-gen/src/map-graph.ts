import { add, compose, pluck, propEq, uniq, without } from 'ramda';
import { Grid, IRoom } from './grid';
import { findNeighborsFrom, weightedPick } from './utils';

export interface IRoomInfo extends IRoom {
  neighbors: Array<IRoomInfo['id']>;
  connections: Array<IRoomInfo['id']>;
}

export const connectRooms = (a: IRoomInfo, b: IRoomInfo) => {
  a.connections = uniq([...a.connections, b.id]);
  b.connections = uniq([...b.connections, a.id]);
};

const updateConnections = (grid: Grid) => {
  const rooms: IRoomInfo[] = [];
  for (const id of grid.listRooms()) {
    const info = grid.findRoomById(id);
    rooms.push({
      ...info,
      neighbors: [],
      connections: [],
    });
  }

  const findNeighbors = findNeighborsFrom(grid.cells);
  // connect rooms that are neighbours
  for (const room of rooms) {
    const ids = [];
    for (const cellId of room.cells) {
      const cell = grid.cells[cellId];
      ids.push(
        ...compose(
          pluck('roomId'),
          findNeighbors,
        )(cell),
      );
    }
    room.neighbors.push(...uniq(ids));
  }

  return rooms;
};

export type MapNode = {
  roomId: number;
  children: MapNode[];
};

export const createGraph = (
  grid: Grid,
  rng?: Pick<Chance.Chance, 'pickone' | 'pickset'>,
) => {
  if (rng === undefined) {
    // tslint:disable-next-line: no-parameter-reassignment
    rng = grid.rng;
  }
  const rooms = updateConnections(grid);
  const rootRoom = rng.pickone(rooms);
  const nodeSize = (node: MapNode): number =>
    node.children.map(nodeSize).reduce(add, 0) + 1;

  const cache = [];

  const asTree = (room: IRoomInfo): MapNode => {
    const childCount = weightedPick([3, 2, 1]);
    // console.log('childCount', childCount);
    const candidates = without(cache, room.neighbors);
    if (candidates.length !== 0) {
      const childIds = rng.pickset(candidates, childCount);
      const childRooms = childIds.map(id => rooms.find(propEq('id', id)));
      for (const child of childRooms) {
        connectRooms(room, child);
      }
    }

    // connections to rooms that haven't been mapped yet
    const children = without(cache, room.connections);

    cache.push(...children);

    return {
      roomId: room.id,
      children: children.map(id => asTree(rooms.find(propEq('id', id)))),
    };
  };

  cache.push(rootRoom.id);

  return asTree(rootRoom);
};
