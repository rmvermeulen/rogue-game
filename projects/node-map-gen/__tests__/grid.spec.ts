import { Grid, IGridOptions } from '../src/grid';
import { times, flip } from 'ramda';

const doN = flip(times);

describe.each([
  {
    width: 5,
    height: 5,
  },
])('grid class', (options: IGridOptions) => {
  const grid = Grid.create(options);

  expect(grid).toBeDefined();

  it('can find rows', () => {
    expect(grid.row(0)).toHaveLength(options.width);
    expect(grid.row(options.height)).toHaveLength(0);
  });

  it('can find columns', () => {
    expect(grid.column(0)).toHaveLength(options.height);
    expect(grid.column(options.width)).toHaveLength(0);
  });

  it('can list the unique room ids', () => {
    expect(grid.listRooms()).toHaveLength(1);
  });

  it('can find cells by position', () => {
    doN(options.width, x =>
      doN(options.height, y =>
        expect(grid.cellAt(x, y)).toMatchObject({ x, y }),
      ),
    );
  });

  it('can be displayed in the cli', () => {
    expect(grid.display()).toMatchSnapshot();
  });
});
