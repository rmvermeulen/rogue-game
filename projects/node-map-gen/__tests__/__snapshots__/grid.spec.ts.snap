// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid class can be displayed in the cli 1`] = `""`;
